# SQL CARTO QUERY FUNCTION

In [1]:
import ast
import pandas as pd

try:  
    from urllib import urlencode
    from urllib2 import HTTPError
    from urllib2 import urlopen
    from StringIO import StringIO as IO

except ImportError:
    from urllib.parse import urlencode
    from urllib.error import HTTPError
    from urllib.request import urlopen, Request
    from io import BytesIO as IO

In [2]:
SQL_SOURCE = 'https://yl5240.carto.com/api/v2/sql?q='
def queryCartoDB(query, format='CSV', source=SQL_SOURCE):
    '''queries carto datasets from a given carto account
    Arguments: 
    query - string: a valid sql query string
    format - outlut format  OPTIONAL (default CSV)
    source - a valid sql api endpoint OPTIONAL (default carto fb55 account)
    Returns:
    the return of the sql query AS A STRING
    NOTES:
    designed for the carto API, tested only with CSV return format'''
    
    data = urlencode({'format': format, 'q': query}).encode("utf-8")
    try:
        response = urlopen(source, data)
    except HTTPError as e:
        raise ValueError('\n'.join(ast.literal_eval(e.readline())['error']))
    except Exception:
        raise
    return response.read()

In [3]:
test = '''
SELECT *
FROM citibike
'''

In [4]:
pd.read_csv(IO(queryCartoDB(test)), sep=',')

the_geom    usertype  the_geom_webmercator  gender  birth_year  bikeid  \
0           NaN  Subscriber                   NaN       1      1965.0   19217   
1           NaN  Subscriber                   NaN       1      1971.0   17575   
2           NaN  Subscriber                   NaN       1      1973.0   20313   
3           NaN  Subscriber                   NaN       2      1978.0   17131   
4           NaN  Subscriber                   NaN       1      1993.0   21289   
5           NaN  Subscriber                   NaN       2      1969.0   18903   
6           NaN  Subscriber                   NaN       2      1985.0   21044   
7           NaN  Subscriber                   NaN       1      1957.0   19868   
8           NaN  Subscriber                   NaN       1      1979.0   15854   
9           NaN  Subscriber                   NaN       1      1983.0   15173   
10          NaN  Subscriber                   NaN       1      1955.0   17862   
11          NaN  Subscriber                   NaN       2      1985.0   21183   
12          NaN  Subscriber                   NaN       1      1982.0   14843   
13          NaN  Subscriber                   NaN       2      1988.0   16936   
14          NaN  Subscriber                   NaN       2      1991.0   21507   
15          NaN  Subscriber                   NaN       2      1982.0   17456   
16          NaN  Subscriber                   NaN       2      1990.0   14679   
17          NaN  Subscriber                   NaN       1      1977.0   18978   
18          NaN  Subscriber                   NaN       1      1968.0   16496   
19          NaN  Subscriber                   NaN       1      1990.0   17107   
20          NaN  Subscriber                   NaN       1      1987.0   16984   
21          NaN  Subscriber                   NaN       1      1980.0   15997   
22          NaN  Subscriber                   NaN       1      1960.0   21516   
23          NaN  Subscriber                   NaN       1      1986.0   15815   
24          NaN  Subscriber                   NaN       1      1987.0   18727   
25          NaN  Subscriber                   NaN       1      1969.0   14982   
26          NaN  Subscriber                   NaN       1      1957.0   16391   
27          NaN  Subscriber                   NaN       1      1981.0   15947   
28          NaN  Subscriber                   NaN       1      1993.0   16917   
29          NaN  Subscriber                   NaN       1      1975.0   15924   
...         ...         ...                   ...     ...         ...     ...   
46170       NaN  Subscriber                   NaN       1      1988.0   15733   
46171       NaN  Subscriber                   NaN       1      1995.0   16931   
46172       NaN  Subscriber                   NaN       1      1987.0   17202   
46173       NaN  Subscriber                   NaN       2      1989.0   20314   
46174       NaN  Subscriber                   NaN       1      1989.0   21531   
46175       NaN  Subscriber                   NaN       1      1968.0   21491   
46176       NaN  Subscriber                   NaN       1      1990.0   15367   
46177       NaN  Subscriber                   NaN       2      1973.0   18978   
46178       NaN  Subscriber                   NaN       1      1980.0   19507   
46179       NaN  Subscriber                   NaN       1      1980.0   16936   
46180       NaN  Subscriber                   NaN       1      1982.0   18689   
46181       NaN  Subscriber                   NaN       1      1972.0   15781   
46182       NaN  Subscriber                   NaN       1      1988.0   15081   
46183       NaN  Subscriber                   NaN       1      1983.0   21672   
46184       NaN  Subscriber                   NaN       1      1970.0   16336   
46185       NaN  Subscriber                   NaN       1      1973.0   21609   
46186       NaN  Subscriber                   NaN       1      1964.0   19470   
46187       NaN  Subscriber           

# Task 1 — Familiarize with SQL Clauses

## Sort data by start_station_id, tripduration 
- Only checking trips with duration <= 3 hours

In [5]:
task1_1 = '''
SELECT * FROM citibike
WHERE tripduration <= 1080
ORDER by start_station_id, tripduration
'''

In [6]:
pd.read_csv(IO(queryCartoDB(task1_1)), sep=',')

the_geom    usertype  the_geom_webmercator  gender  birth_year  bikeid  \
0           NaN  Subscriber                   NaN       2      1980.0   19039   
1           NaN  Subscriber                   NaN       2      1959.0   18131   
2           NaN  Subscriber                   NaN       1      1995.0   19729   
3           NaN  Subscriber                   NaN       1      1970.0   16879   
4           NaN  Subscriber                   NaN       1      1983.0   15277   
5           NaN  Subscriber                   NaN       1      1971.0   18492   
6           NaN  Subscriber                   NaN       1      1961.0   19216   
7           NaN  Subscriber                   NaN       1      1974.0   21588   
8           NaN  Subscriber                   NaN       1      1978.0   19124   
9           NaN  Subscriber                   NaN       1      1974.0   18174   
10          NaN  Subscriber                   NaN       1      1983.0   18198   
11          NaN  Subscriber                   NaN       1      1982.0   19672   
12          NaN  Subscriber                   NaN       2      1986.0   14692   
13          NaN  Subscriber                   NaN       1      1983.0   17669   
14          NaN  Subscriber                   NaN       1      1961.0   16754   
15          NaN  Subscriber                   NaN       1      1983.0   21114   
16          NaN  Subscriber                   NaN       1      1970.0   19729   
17          NaN  Subscriber                   NaN       1      1981.0   18620   
18          NaN  Subscriber                   NaN       1      1981.0   20119   
19          NaN  Subscriber                   NaN       1      1981.0   20988   
20          NaN  Subscriber                   NaN       2      1972.0   16310   
21          NaN  Subscriber                   NaN       1      1978.0   18287   
22          NaN  Subscriber                   NaN       1      1973.0   18991   
23          NaN  Subscriber                   NaN       1      1981.0   20600   
24          NaN  Subscriber                   NaN       1      1973.0   15927   
25          NaN  Subscriber                   NaN       1      1991.0   15282   
26          NaN  Subscriber                   NaN       1      1978.0   16405   
27          NaN  Subscriber                   NaN       1      1980.0   17196   
28          NaN  Subscriber                   NaN       2      1986.0   17972   
29          NaN  Subscriber                   NaN       1      1974.0   19771   
...         ...         ...                   ...     ...         ...     ...   
40282       NaN  Subscriber                   NaN       2      1980.0   16987   
40283       NaN  Subscriber                   NaN       1      1996.0   19309   
40284       NaN  Subscriber                   NaN       1      1977.0   15189   
40285       NaN  Subscriber                   NaN       1      1944.0   15510   
40286       NaN    Customer                   NaN       0         NaN   17899   
40287       NaN  Subscriber                   NaN       1      1960.0   19303   
40288       NaN  Subscriber                   NaN       1      1980.0   20901   
40289       NaN  Subscriber                   NaN       2      1967.0   18386   
40290       NaN  Subscriber                   NaN       1      1991.0   19783   
40291       NaN  Subscriber                   NaN       1      1951.0   15977   
40292       NaN  Subscriber                   NaN       2      1985.0   15977   
40293       NaN  Subscriber                   NaN       1      1979.0   17004   
40294       NaN  Subscriber                   NaN       1      1959.0   19434   
40295       NaN  Subscriber                   NaN       1      1972.0   20637   
40296       NaN  Subscriber                   NaN       1      1956.0   17180   
40297       NaN  Subscriber                   NaN       1      1977.0   21407   
40298       NaN  Subscriber                   NaN       1      1959.0   15748   
40299       NaN  Subscriber           

## Only show the top/last 10 records (aka head and tail in SQL) 

In [7]:
task1_2 = '''
SELECT * FROM citibike
WHERE tripduration <= 1080
ORDER by start_station_id, tripduration
LIMIT 10
'''
pd.read_csv(IO(queryCartoDB(task1_2)), sep=',')

the_geom    usertype  the_geom_webmercator  gender  birth_year  bikeid  \
0       NaN  Subscriber                   NaN       2        1980   19039   
1       NaN  Subscriber                   NaN       2        1959   18131   
2       NaN  Subscriber                   NaN       1        1995   19729   
3       NaN  Subscriber                   NaN       1        1970   16879   
4       NaN  Subscriber                   NaN       1        1983   15277   
5       NaN  Subscriber                   NaN       1        1971   18492   
6       NaN  Subscriber                   NaN       1        1961   19216   
7       NaN  Subscriber                   NaN       1        1974   21588   
8       NaN  Subscriber                   NaN       1        1978   19124   
9       NaN  Subscriber                   NaN       1        1974   18174   

   end_station_id  start_station_id  tripduration  cartodb_id  \
0              72                72           107        8998   
1              72                72           189       40759   
2              72                72           229       40757   
3             478                72           248       21948   
4             515                72           252       43099   
5             478                72           261       32566   
6             478                72           276       15971   
7             447                72           283       41355   
8             514                72           288       42074   
9             447                72           301        2282   

   end_station_name start_station_name  end_station_longitude  \
0  W 52 St & 11 Ave   W 52 St & 11 Ave             -73.993929   
1  W 52 St & 11 Ave   W 52 St & 11 Ave             -73.993929   
2  W 52 St & 11 Ave   W 52 St & 11 Ave             -73.993929   
3  11 Ave & W 41 St   W 52 St & 11 Ave             -73.998842   
4  W 43 St & 10 Ave   W 52 St & 11 Ave             -73.994618   
5  11 Ave & W 41 St   W 52 St & 11 Ave             -73.998842   
6  11 Ave & W 41 St   W 52 St & 11 Ave             -73.998842   
7   8 Ave & W 52 St   W 52 St & 11 Ave             -73.985162   
8  12 Ave & W 40 St   W 52 St & 11 Ave             -74.002777   
9   8 Ave & W 52 St   W 52 St & 11 Ave             -73.985162   

   end_station_latitude  start_station_longitude  start_station_latitude  \
0             40.767272               -73.993929               40.767272   
1             40.767272               -73.993929               40.767272   
2             40.767272               -73.993929               40.767272   
3             40.760301               -73.993929               40.767272   
4             40.760094               -73.993929               40.767272   
5             40.760301               -73.993929               40.767272   
6             40.760301               -73.993929               40.767272   
7             40.763707               -73.993929               40.767272   
8             40.760875               -73.993929               40.767272   
9             40.763707               -73.993929               40.767272   

                 stoptime               starttime  
0  2015-02-03 10:12:00+00  2015-02-03 10:10:00+00  
1  2015-02-07 10:43:00+00  2015-02-07 10:40:00+00  
2  2015-02-07 10:44:00+00  2015-02-07 10:40:00+00  
3  2015-02-05 06:24:00+00  2015-02-05 06:19:00+00  
4  2015-02-07 15:47:00+00  2015-02-07 15:43:00+00  
5  2015-02-06 09:02:00+00  2015-02-06 08:58:00+00  
6  2015-02-04 12:51:00+00  2015-02-04 12:46:00+00  
7  2015-02-07 12:15:00+00  2015-02-07 12:10:00+00  
8  2015-02-07 13:54:00+00  2015-02-07 13:50:00+00  
9  2015-02-01 13:13:00+00  2015-02-01 13:08:00+00

In [8]:
task1_3 = '''
SELECT * FROM citibike
WHERE tripduration <= 1080
ORDER by start_station_id DESC, tripduration DESC
LIMIT 10
'''
pd.read_csv(IO(queryCartoDB(task1_3)), sep=',')

the_geom    usertype  the_geom_webmercator  gender  birth_year  bikeid  \
0       NaN  Subscriber                   NaN       2      1973.0   18797   
1       NaN  Subscriber                   NaN       1      1998.0   15090   
2       NaN  Subscriber                   NaN       2      1973.0   14906   
3       NaN  Subscriber                   NaN       1      1997.0   17310   
4       NaN    Customer                   NaN       0         NaN   16565   
5       NaN  Subscriber                   NaN       1      1983.0   20379   
6       NaN  Subscriber                   NaN       2      1964.0   20744   
7       NaN  Subscriber                   NaN       2      1980.0   20156   
8       NaN  Subscriber                   NaN       2      1973.0   16144   
9       NaN  Subscriber                   NaN       1      1988.0   15904   

   end_station_id  start_station_id  tripduration  cartodb_id  \
0             401              3002          1058       28070   
1             128              3002          1045       38836   
2             401              3002          1032       20648   
3             128              3002          1031       38837   
4             128              3002          1005       44156   
5             496              3002           951       40770   
6             531              3002           943        4413   
7             263              3002           938       36626   
8             401              3002           931       12373   
9            2000              3002           925       15488   

           end_station_name          start_station_name  \
0   Allen St & Rivington St  South End Ave & Liberty St   
1  MacDougal St & Prince St  South End Ave & Liberty St   
2   Allen St & Rivington St  South End Ave & Liberty St   
3  MacDougal St & Prince St  South End Ave & Liberty St   
4  MacDougal St & Prince St  South End Ave & Liberty St   
5           E 16 St & 5 Ave  South End Ave & Liberty St   
6    Forsyth St & Broome St  South End Ave & Liberty St   
7  Elizabeth St & Hester St  South End Ave & Liberty St   
8   Allen St & Rivington St  South End Ave & Liberty St   
9  Front St & Washington St  South End Ave & Liberty St   

   end_station_longitude  end_station_latitude  start_station_longitude  \
0             -73.989978             40.720196               -74.015756   
1             -74.002971             40.727103               -74.015756   
2             -73.989978             40.720196               -74.015756   
3             -74.002971             40.727103               -74.015756   
4             -74.002971             40.727103               -74.015756   
5             -73.992390             40.737262               -74.015756   
6             -73.992663             40.718939               -74.015756   
7             -73.996375             40.717290               -74.015756   
8             -73.989978             40.720196               -74.015756   
9             -73.989402             40.702551               -74.015756   

   start_station_latitude                stoptime               starttime  
0               40.711512  2015-02-05 17:42:00+00  2015-02-05 17:24:00+00  
1               40.711512  2015-02-06 20:57:00+00  2015-02-06 20:40:00+00  
2               40.711512  2015-02-04 20:41:00+00  2015-02-04 20:24:00+00  
3               40.711512  2015-02-06 20:57:00+00  2015-02-06 20:40:00+00  
4               40.711512  2015-02-07 17:45:00+00  2015-02-07 17:29:00+00  
5               40.711512  2015-02-07 10:58:00+00  2015-02-07 10:42:00+00  
6               40.711512  2015-02-01 17:00:00+00  2015-02-01 16:45:00+00  
7               40.711512  2015-02-06 17:40:00+00  2015-02-06 17:25:00+00  
8               40.711512  2015-02-03 21:51:00+00  2015-02-03 21:35:00+00  
9               40.711512  2015-02-04 11:40:00+00  2015-02-04 11:25:00+00

## List all unique start_station_id values 

In [9]:
task1_4 = '''
SELECT DISTINCT start_station_id FROM citibike
ORDER BY start_station_id
'''
pd.read_csv(IO(queryCartoDB(task1_4)), sep=',')

start_station_id  Unnamed: 1
0                  72         NaN
1                  79         NaN
2                  82         NaN
3                  83         NaN
4                 116         NaN
5                 119         NaN
6                 120         NaN
7                 127         NaN
8                 128         NaN
9                 137         NaN
10                143         NaN
11                144         NaN
12                146         NaN
13                147         NaN
14                150         NaN
15                151         NaN
16                152         NaN
17                153         NaN
18                157         NaN
19                160         NaN
20                161         NaN
21                164         NaN
22                167         NaN
23                168         NaN
24                173         NaN
25                174         NaN
26                195         NaN
27                212         NaN
28                216         NaN
29                217         NaN
..                ...         ...
297               525         NaN
298               526         NaN
299               527         NaN
300               528         NaN
301               529         NaN
302               530         NaN
303               531         NaN
304               532         NaN
305               533         NaN
306               534         NaN
307               537         NaN
308               538         NaN
309               539         NaN
310               545         NaN
311               546         NaN
312              2000         NaN
313              2001         NaN
314              2002         NaN
315              2004         NaN
316              2005         NaN
317              2006         NaN
318              2008         NaN
319              2009         NaN
320              2010         NaN
321              2012         NaN
322              2017         NaN
323              2021         NaN
324              2022         NaN
325              2023         NaN
326              3002         NaN

[327 rows x 2 columns]

## Aggregation functions: 
- Count the number of trips (aka wc -l in SQL) 
-  Find the average/min/max trip duration

In [10]:
task1_5 = '''
SELECT COUNT(cartodb_id) FROM citibike
'''
pd.read_csv(IO(queryCartoDB(task1_5)), sep=',')

count  Unnamed: 1
0  46200         NaN

In [11]:
task1_6 = '''
SELECT AVG(tripduration),MIN(tripduration),MAX(tripduration) FROM citibike
'''
pd.read_csv(IO(queryCartoDB(task1_6)), sep=',')

avg  min    max
0  675.865823   60  43016

# Task 2 — Working with date/time

## Selecting trips started on Feb-02-2015 only 

In [12]:
task2_1 = '''
SELECT * FROM citibike
WHERE starttime <= '2015-02-02 23:59:59' AND starttime > '2015-02-01 23:59:59'
'''
pd.read_csv(IO(queryCartoDB(task2_1)), sep=',')

the_geom    usertype  the_geom_webmercator  gender  birth_year  bikeid  \
0          NaN  Subscriber                   NaN       1      1992.0   20684   
1          NaN  Subscriber                   NaN       2      1964.0   16094   
2          NaN  Subscriber                   NaN       1      1974.0   17362   
3          NaN  Subscriber                   NaN       2      1992.0   15475   
4          NaN  Subscriber                   NaN       1      1968.0   16272   
5          NaN  Subscriber                   NaN       1      1979.0   16370   
6          NaN  Subscriber                   NaN       1      1963.0   19437   
7          NaN  Subscriber                   NaN       1      1982.0   21211   
8          NaN  Subscriber                   NaN       2      1978.0   21417   
9          NaN  Subscriber                   NaN       1      1956.0   15722   
10         NaN  Subscriber                   NaN       1      1962.0   18972   
11         NaN  Subscriber                   NaN       1      1980.0   19628   
12         NaN  Subscriber                   NaN       1      1978.0   15169   
13         NaN  Subscriber                   NaN       1      1997.0   17533   
14         NaN  Subscriber                   NaN       1      1983.0   19761   
15         NaN  Subscriber                   NaN       1      1993.0   17267   
16         NaN  Subscriber                   NaN       1      1964.0   18691   
17         NaN  Subscriber                   NaN       1      1977.0   17106   
18         NaN  Subscriber                   NaN       1      1973.0   20009   
19         NaN  Subscriber                   NaN       1      1979.0   18777   
20         NaN  Subscriber                   NaN       1      1989.0   15357   
21         NaN  Subscriber                   NaN       1      1983.0   18487   
22         NaN  Subscriber                   NaN       1      1981.0   20863   
23         NaN  Subscriber                   NaN       1      1970.0   16539   
24         NaN  Subscriber                   NaN       1      1961.0   18737   
25         NaN  Subscriber                   NaN       2      1980.0   17797   
26         NaN  Subscriber                   NaN       1      1990.0   20638   
27         NaN  Subscriber                   NaN       1      1989.0   17636   
28         NaN  Subscriber                   NaN       1      1982.0   18472   
29         NaN  Subscriber                   NaN       1      1968.0   17456   
...        ...         ...                   ...     ...         ...     ...   
1429       NaN  Subscriber                   NaN       1      1985.0   20572   
1430       NaN  Subscriber                   NaN       1      1981.0   20338   
1431       NaN  Subscriber                   NaN       1      1978.0   15826   
1432       NaN  Subscriber                   NaN       1      1990.0   18325   
1433       NaN  Subscriber                   NaN       1      1977.0   16654   
1434       NaN  Subscriber                   NaN       1      1966.0   20902   
1435       NaN  Subscriber                   NaN       1      1954.0   20796   
1436       NaN  Subscriber                   NaN       1      1989.0   20160   
1437       NaN  Subscriber                   NaN       1      1974.0   17392   
1438       NaN  Subscriber                   NaN       1      1963.0   17263   
1439       NaN  Subscriber                   NaN       1      1955.0   19858   
1440       NaN  Subscriber                   NaN       1      1979.0   20319   
1441       NaN  Subscriber                   NaN       1      1977.0   21600   
1442       NaN  Subscriber                   NaN       1      1986.0   18791   
1443       NaN  Subscriber                   NaN       1      1966.0   17730   
1444       NaN  Subscriber                   NaN       1      1980.0   18147   
1445       NaN  Subscriber                   NaN       1      1959.0   20249   
1446       NaN  Subscriber                   NaN       1      1988.0   18078   
1447  

## Selecting trips started on the weekends
- What are average trip duration during weekends? 

In [13]:
task2_2 = '''
SELECT * FROM citibike
WHERE extract(DOW FROM starttime) in (0,6) 
'''
pd.read_csv(IO(queryCartoDB(task2_2)), sep=',')

the_geom    usertype  the_geom_webmercator  gender  birth_year  bikeid  \
0           NaN  Subscriber                   NaN       1      1965.0   19217   
1           NaN  Subscriber                   NaN       2      1978.0   17131   
2           NaN  Subscriber                   NaN       1      1993.0   21289   
3           NaN  Subscriber                   NaN       2      1969.0   18903   
4           NaN  Subscriber                   NaN       2      1985.0   21044   
5           NaN  Subscriber                   NaN       1      1957.0   19868   
6           NaN  Subscriber                   NaN       1      1979.0   15854   
7           NaN  Subscriber                   NaN       1      1983.0   15173   
8           NaN  Subscriber                   NaN       1      1955.0   17862   
9           NaN  Subscriber                   NaN       2      1985.0   21183   
10          NaN  Subscriber                   NaN       1      1982.0   14843   
11          NaN  Subscriber                   NaN       2      1988.0   16936   
12          NaN  Subscriber                   NaN       2      1991.0   21507   
13          NaN  Subscriber                   NaN       2      1982.0   17456   
14          NaN  Subscriber                   NaN       2      1990.0   14679   
15          NaN  Subscriber                   NaN       1      1977.0   18978   
16          NaN  Subscriber                   NaN       1      1968.0   16496   
17          NaN  Subscriber                   NaN       1      1990.0   17107   
18          NaN  Subscriber                   NaN       1      1987.0   16984   
19          NaN  Subscriber                   NaN       1      1980.0   15997   
20          NaN  Subscriber                   NaN       1      1960.0   21516   
21          NaN  Subscriber                   NaN       1      1986.0   15815   
22          NaN  Subscriber                   NaN       1      1987.0   18727   
23          NaN  Subscriber                   NaN       1      1969.0   14982   
24          NaN  Subscriber                   NaN       1      1957.0   16391   
25          NaN  Subscriber                   NaN       1      1981.0   15947   
26          NaN  Subscriber                   NaN       1      1993.0   16917   
27          NaN  Subscriber                   NaN       1      1975.0   15924   
28          NaN  Subscriber                   NaN       1      1973.0   21321   
29          NaN  Subscriber                   NaN       2      1973.0   16861   
...         ...         ...                   ...     ...         ...     ...   
13201       NaN  Subscriber                   NaN       1      1988.0   15733   
13202       NaN  Subscriber                   NaN       1      1995.0   16931   
13203       NaN  Subscriber                   NaN       1      1987.0   17202   
13204       NaN  Subscriber                   NaN       2      1989.0   20314   
13205       NaN  Subscriber                   NaN       1      1989.0   21531   
13206       NaN  Subscriber                   NaN       1      1968.0   21491   
13207       NaN  Subscriber                   NaN       1      1990.0   15367   
13208       NaN  Subscriber                   NaN       2      1973.0   18978   
13209       NaN  Subscriber                   NaN       1      1980.0   19507   
13210       NaN  Subscriber                   NaN       1      1980.0   16936   
13211       NaN  Subscriber                   NaN       1      1982.0   18689   
13212       NaN  Subscriber                   NaN       1      1972.0   15781   
13213       NaN  Subscriber                   NaN       1      1988.0   15081   
13214       NaN  Subscriber                   NaN       1      1983.0   21672   
13215       NaN  Subscriber                   NaN       1      1970.0   16336   
13216       NaN  Subscriber                   NaN       1      1973.0   21609   
13217       NaN  Subscriber                   NaN       1      1964.0   19470   
13218       NaN  Subscriber           

In [14]:
task2_3 = '''
SELECT AVG(tripduration) as avg_weekends FROM citibike
WHERE extract(DOW FROM starttime) in (0,6) 
'''
pd.read_csv(IO(queryCartoDB(task2_3)), sep=',')

avg_weekends  Unnamed: 1
0    662.942181         NaN

### The average trip duration during weekends is 663 seconds, i.e. 10.05 mins.

## Selecting trips started on the weekdays
- What are average trip duration during weekdays? 

In [15]:
task2_4 = '''
SELECT * FROM citibike
WHERE extract(DOW FROM starttime) NOT IN (0,6) 
'''
pd.read_csv(IO(queryCartoDB(task2_4)), sep=',')

the_geom    usertype  the_geom_webmercator  gender  birth_year  bikeid  \
0           NaN  Subscriber                   NaN       1      1971.0   17575   
1           NaN  Subscriber                   NaN       1      1973.0   20313   
2           NaN  Subscriber                   NaN       1      1992.0   20684   
3           NaN  Subscriber                   NaN       2      1964.0   16094   
4           NaN  Subscriber                   NaN       1      1974.0   17362   
5           NaN  Subscriber                   NaN       2      1992.0   15475   
6           NaN  Subscriber                   NaN       1      1968.0   16272   
7           NaN  Subscriber                   NaN       1      1979.0   16370   
8           NaN  Subscriber                   NaN       1      1963.0   19437   
9           NaN  Subscriber                   NaN       1      1982.0   21211   
10          NaN  Subscriber                   NaN       2      1978.0   21417   
11          NaN  Subscriber                   NaN       1      1956.0   15722   
12          NaN  Subscriber                   NaN       1      1962.0   18972   
13          NaN  Subscriber                   NaN       1      1980.0   19628   
14          NaN  Subscriber                   NaN       1      1978.0   15169   
15          NaN  Subscriber                   NaN       1      1997.0   17533   
16          NaN  Subscriber                   NaN       1      1983.0   19761   
17          NaN  Subscriber                   NaN       1      1993.0   17267   
18          NaN  Subscriber                   NaN       1      1964.0   18691   
19          NaN  Subscriber                   NaN       1      1977.0   17106   
20          NaN  Subscriber                   NaN       1      1973.0   20009   
21          NaN  Subscriber                   NaN       1      1979.0   18777   
22          NaN  Subscriber                   NaN       1      1989.0   15357   
23          NaN  Subscriber                   NaN       1      1983.0   18487   
24          NaN  Subscriber                   NaN       1      1981.0   20863   
25          NaN  Subscriber                   NaN       1      1970.0   16539   
26          NaN  Subscriber                   NaN       1      1961.0   18737   
27          NaN  Subscriber                   NaN       2      1980.0   17797   
28          NaN  Subscriber                   NaN       1      1990.0   20638   
29          NaN  Subscriber                   NaN       1      1989.0   17636   
...         ...         ...                   ...     ...         ...     ...   
32939       NaN  Subscriber                   NaN       1      1957.0   18249   
32940       NaN  Subscriber                   NaN       2      1972.0   19146   
32941       NaN  Subscriber                   NaN       1      1984.0   15313   
32942       NaN  Subscriber                   NaN       1      1979.0   18469   
32943       NaN  Subscriber                   NaN       1      1966.0   18929   
32944       NaN  Subscriber                   NaN       1      1987.0   19009   
32945       NaN  Subscriber                   NaN       1      1984.0   17102   
32946       NaN  Subscriber                   NaN       2      1972.0   18686   
32947       NaN  Subscriber                   NaN       1      1989.0   15377   
32948       NaN  Subscriber                   NaN       1      1984.0   17844   
32949       NaN  Subscriber                   NaN       1      1987.0   21040   
32950       NaN  Subscriber                   NaN       1      1981.0   19480   
32951       NaN  Subscriber                   NaN       1      1978.0   21130   
32952       NaN  Subscriber                   NaN       1      1983.0   16934   
32953       NaN  Subscriber                   NaN       1      1968.0   20745   
32954       NaN  Subscriber                   NaN       1      1969.0   16811   
32955       NaN  Subscriber                   NaN       1      1980.0   21259   
32956       NaN  Subscriber           

In [16]:
task2_5 = '''
SELECT AVG(tripduration) as avg_weekdays FROM citibike
WHERE extract(DOW FROM starttime) NOT IN (0,6) 
'''
pd.read_csv(IO(queryCartoDB(task2_5)), sep=',')

avg_weekdays  Unnamed: 1
0    681.052292         NaN

### The average trip duration during weekdays is 681 seconds, i.e. 11.35 mins.

# Task 3 — Working with Space

## Showing the list of start station locations
- Using GROUP BY

In [17]:
task3_1 = '''
SELECT CDB_TransformToWebmercator(CDB_LatLng(start_station_latitude, 
                                start_station_longitude)) as the_geom_webmercator,
       MIN(cartodb_id) as cartodb_id,
       MIN(start_station_name) as start_station_name 
FROM citibike
GROUP BY start_station_id, start_station_latitude, start_station_longitude
ORDER BY cartodb_id
'''
pd.read_csv(IO(queryCartoDB(task3_1)), sep=',')

the_geom_webmercator  cartodb_id  \
0    0101000020110F000042FA2528026C5FC1F9B9029FFCFA...           1   
1    0101000020110F0000B72236DE7E6B5FC1A5B36C6610F9...           2   
2    0101000020110F0000309ECD0FB7695FC14F41EB09FDFC...           3   
3    0101000020110F0000262DCC7D156D5FC111F247CB3FF7...           4   
4    0101000020110F0000833ADCCB156B5FC1BEFBAE9CA7F2...           5   
5    0101000020110F0000C12D0F658D675FC1323CC0ADCAF2...           6   
6    0101000020110F00006B6C9CBF186A5FC109C57CA0D8FC...           7   
7    0101000020110F00005A26772B766B5FC11C6CFDE484F7...           8   
8    0101000020110F0000F993EDE4EC6B5FC122E50B32EAF7...           9   
9    0101000020110F00003D22E9FBDB6C5FC18DD337A658F9...          10   
10   0101000020110F00002E1B9FF6466B5FC1E99EE4C5B6FA...          12   
11   0101000020110F0000049C1A69CE6C5FC1205139A81BF8...          13   
12   0101000020110F0000A87451BDDC6A5FC1EB7B4387FCF7...          14   
13   0101000020110F0000517121EA906B5FC182C0396DB4F8...          15   
14   0101000020110F0000047C4EA0256A5FC1D9059AB05DF7...          17   
15   0101000020110F00009BB87D02B76A5FC146C144E9E5FB...          19   
16   0101000020110F0000C06D97808D6A5FC1BD39524AAEF7...          20   
17   0101000020110F000091C14B01406B5FC1A3DCAD024AF7...          21   
18   0101000020110F00009F5AD0F883695FC18BD6351F95FB...          22   
19   0101000020110F0000E5C68C67566B5FC1CFC9ACC28AFD...          23   
20   0101000020110F0000303622AF5E6A5FC1A3EB11A85CF9...          24   
21   0101000020110F0000EC774642E56C5FC13DF29213A3F7...          26   
22   0101000020110F00002EBB7FC3E96B5FC13578821125FB...          27   
23   0101000020110F00001E76C09C296B5FC16ACD1485FCF5...          28   
24   0101000020110F0000EEE2EC99196C5FC100162DBD45F7...          29   
25   0101000020110F0000505A780E946A5FC1B7C653F55EF8...          30   
26   0101000020110F00001154E295996A5FC10680C9DF44F7...          31   
27   0101000020110F000042517D0A9F6A5FC15138BD50E5F8...          32   
28   0101000020110F0000F55695027D6B5FC186E5BB69D7FB...          33   
29   0101000020110F0000DCAA7863C66B5FC11BA5784D77F6...          35   
..                                                 ...         ...   
297  0101000020110F000019F7C7B2096A5FC1F10194C3E0FB...        2433   
298  0101000020110F0000D387C368DE685FC193F25C8393FB...        2462   
299  0101000020110F0000E463F723266A5FC14F49CEE530FB...        2481   
300  0101000020110F00000EDB1043D86A5FC1098E90170AF2...        2509   
301  0101000020110F0000BAF51EEBD8695FC1C4553BF749F2...        2558   
302  0101000020110F000099F9690747695FC1D617A37CFAF1...        2577   
303  0101000020110F00008077AE3304685FC1A3E1631FF7F2...        2762   
304  0101000020110F0000F2F00227D6685FC1E3F0607ED0F1...        2941   
305  0101000020110F0000DA5401006A6A5FC1177BC9DEA3FB...        3167   
306  0101000020110F0000B4B6D9A8FF695FC1B42BC0556EF1...        3237   
307  0101000020110F00009A6669876B695FC10D48A3FB82F4...        3353   
308  0101000020110F0000E2BF6C8A0D6E5FC1932920126AF4...        3617   
309  0101000020110F0000BED1C0F782685FC1A805EBBA8EF5...        3645   
310  0101000020110F00001C8FDEBEA46B5FC1F99BE632D1F3...        3700   
311  0101000020110F00005CB2E76C746B5FC16FC1B8F4BCF3...        3902   
312  0101000020110F00007D0BEFFB916C5FC19358D78D8DF2...        4252   
313  0101000020110F0000C5124F78DA695FC19387858CA7FB...        4267   
314  0101000020110F000013E67A93D76D5FC1E085DF0946F4...        4308   
315  0101000020110F0000A8AED80FD6675FC123D0DAC368F2...        4544   
316  0101000020110F000097C9A809E76B5FC15DF6A6B925F4...        4666   
317  0101000020110F000052716DFB7B6D5FC1634AEE724BF4...        5444   
318  0101000020110F000029BF3029796A5FC13B53193685F3...        5816   
319  0101000020110F00006F1B0D7971695FC1AA39F250F4FA...        7030   
320  0101000020110F00007B70CCAF376B5FC1099854822BFB...        7544   
321  0101000020110F0000262E4F4BAF6A5FC18F9D963E1FF2...        7628   
322  0101000020110F0000C8A0F7054C695FC137B64

## Showing the number of trips started per station 

In [18]:
task3_2 = '''
SELECT CDB_TransformToWebmercator(CDB_LatLng(start_station_latitude, 
                        start_station_longitude)) as the_geom_webmercator,
       MIN(cartodb_id) as cartodb_id,
       MIN(start_station_name) as start_station_name,
       COUNT(cartodb_id) as number_trips
FROM citibike
GROUP BY start_station_id, start_station_latitude, start_station_longitude
ORDER BY cartodb_id
'''
pd.read_csv(IO(queryCartoDB(task3_2)), sep=',')

the_geom_webmercator  cartodb_id  \
0    0101000020110F000042FA2528026C5FC1F9B9029FFCFA...           1   
1    0101000020110F0000B72236DE7E6B5FC1A5B36C6610F9...           2   
2    0101000020110F0000309ECD0FB7695FC14F41EB09FDFC...           3   
3    0101000020110F0000262DCC7D156D5FC111F247CB3FF7...           4   
4    0101000020110F0000833ADCCB156B5FC1BEFBAE9CA7F2...           5   
5    0101000020110F0000C12D0F658D675FC1323CC0ADCAF2...           6   
6    0101000020110F00006B6C9CBF186A5FC109C57CA0D8FC...           7   
7    0101000020110F00005A26772B766B5FC11C6CFDE484F7...           8   
8    0101000020110F0000F993EDE4EC6B5FC122E50B32EAF7...           9   
9    0101000020110F00003D22E9FBDB6C5FC18DD337A658F9...          10   
10   0101000020110F00002E1B9FF6466B5FC1E99EE4C5B6FA...          12   
11   0101000020110F0000049C1A69CE6C5FC1205139A81BF8...          13   
12   0101000020110F0000A87451BDDC6A5FC1EB7B4387FCF7...          14   
13   0101000020110F0000517121EA906B5FC182C0396DB4F8...          15   
14   0101000020110F0000047C4EA0256A5FC1D9059AB05DF7...          17   
15   0101000020110F00009BB87D02B76A5FC146C144E9E5FB...          19   
16   0101000020110F0000C06D97808D6A5FC1BD39524AAEF7...          20   
17   0101000020110F000091C14B01406B5FC1A3DCAD024AF7...          21   
18   0101000020110F00009F5AD0F883695FC18BD6351F95FB...          22   
19   0101000020110F0000E5C68C67566B5FC1CFC9ACC28AFD...          23   
20   0101000020110F0000303622AF5E6A5FC1A3EB11A85CF9...          24   
21   0101000020110F0000EC774642E56C5FC13DF29213A3F7...          26   
22   0101000020110F00002EBB7FC3E96B5FC13578821125FB...          27   
23   0101000020110F00001E76C09C296B5FC16ACD1485FCF5...          28   
24   0101000020110F0000EEE2EC99196C5FC100162DBD45F7...          29   
25   0101000020110F0000505A780E946A5FC1B7C653F55EF8...          30   
26   0101000020110F00001154E295996A5FC10680C9DF44F7...          31   
27   0101000020110F000042517D0A9F6A5FC15138BD50E5F8...          32   
28   0101000020110F0000F55695027D6B5FC186E5BB69D7FB...          33   
29   0101000020110F0000DCAA7863C66B5FC11BA5784D77F6...          35   
..                                                 ...         ...   
297  0101000020110F000019F7C7B2096A5FC1F10194C3E0FB...        2433   
298  0101000020110F0000D387C368DE685FC193F25C8393FB...        2462   
299  0101000020110F0000E463F723266A5FC14F49CEE530FB...        2481   
300  0101000020110F00000EDB1043D86A5FC1098E90170AF2...        2509   
301  0101000020110F0000BAF51EEBD8695FC1C4553BF749F2...        2558   
302  0101000020110F000099F9690747695FC1D617A37CFAF1...        2577   
303  0101000020110F00008077AE3304685FC1A3E1631FF7F2...        2762   
304  0101000020110F0000F2F00227D6685FC1E3F0607ED0F1...        2941   
305  0101000020110F0000DA5401006A6A5FC1177BC9DEA3FB...        3167   
306  0101000020110F0000B4B6D9A8FF695FC1B42BC0556EF1...        3237   
307  0101000020110F00009A6669876B695FC10D48A3FB82F4...        3353   
308  0101000020110F0000E2BF6C8A0D6E5FC1932920126AF4...        3617   
309  0101000020110F0000BED1C0F782685FC1A805EBBA8EF5...        3645   
310  0101000020110F00001C8FDEBEA46B5FC1F99BE632D1F3...        3700   
311  0101000020110F00005CB2E76C746B5FC16FC1B8F4BCF3...        3902   
312  0101000020110F00007D0BEFFB916C5FC19358D78D8DF2...        4252   
313  0101000020110F0000C5124F78DA695FC19387858CA7FB...        4267   
314  0101000020110F000013E67A93D76D5FC1E085DF0946F4...        4308   
315  0101000020110F0000A8AED80FD6675FC123D0DAC368F2...        4544   
316  0101000020110F000097C9A809E76B5FC15DF6A6B925F4...        4666   
317  0101000020110F000052716DFB7B6D5FC1634AEE724BF4...        5444   
318  0101000020110F000029BF3029796A5FC13B53193685F3...        5816   
319  0101000020110F00006F1B0D7971695FC1AA39F250F4FA...        7030   
320  0101000020110F00007B70CCAF376B5FC1099854822BFB...        7544   
321  0101000020110F0000262E4F4BAF6A5FC18F9D963E1FF2...        7628   
322  0101000020110F0000C8A0F7054C695FC137B64

# Showing the number of trips started per station only for stations within 500m of Time Square! 
- The coordinates of Time Square is (40.7577,-73.9857)

In [19]:
task3_3 = '''
SELECT CDB_TransformToWebmercator(CDB_LatLng(start_station_latitude, 
                                start_station_longitude)) as the_geom_webmercator,
       MIN(cartodb_id) as cartodb_id,
       MIN(start_station_name) as start_station_name,
       COUNT(cartodb_id) as number_trips
FROM citibike
WHERE ST_DWithin(CDB_LatLng(start_station_latitude, 
                    start_station_longitude)::geography,
                    CDB_LatLng(40.7577,-73.9857)::geography,500)
GROUP BY start_station_id, start_station_latitude, start_station_longitude
ORDER BY cartodb_id
'''
pd.read_csv(IO(queryCartoDB(task3_3)), sep=',')

the_geom_webmercator  cartodb_id  \
0  0101000020110F00009BB87D02B76A5FC146C144E9E5FB...          19   
1  0101000020110F0000F55695027D6B5FC186E5BB69D7FB...          33   
2  0101000020110F000020D6742CE16A5FC112F4D73D73FC...          42   
3  0101000020110F000004AF971C1D6B5FC17A48F3C1A8FB...          54   
4  0101000020110F00001607D538556B5FC119A6CA6F41FC...         124   
5  0101000020110F00003F9300B2976B5FC1B8B88F3102FC...         783   
6  0101000020110F0000F22A632FBE6A5FC1363A28CFADFB...        1115   

   start_station_name  number_trips  
0     W 45 St & 6 Ave           141  
1     W 41 St & 8 Ave           507  
2  Broadway & W 49 St           213  
3  Broadway & W 41 St           251  
4     W 45 St & 8 Ave           141  
5     W 42 St & 8 Ave           221  
6     W 43 St & 6 Ave           112

# Task 4 — Putting it all together

## Find the station that had the longest average trip duration during weekends and within 500m of TimeSquare! 

In [20]:
task4_1 = '''
SELECT CDB_TransformToWebmercator(CDB_LatLng(start_station_latitude, 
                                start_station_longitude)) as the_geom_webmercator,
       MIN(cartodb_id) as cartodb_id,
       MIN(start_station_name) as start_station_name,
       AVG(tripduration) as average_trip_duration,
       COUNT(cartodb_id) as number_trips
FROM citibike
WHERE ST_DWithin(CDB_LatLng(start_station_latitude, 
                    start_station_longitude)::geography,
                    CDB_LatLng(40.7577,-73.9857)::geography,500) 
                    AND extract(DOW FROM starttime) in (0, 6)
GROUP BY start_station_id, start_station_latitude, start_station_longitude
ORDER BY average_trip_duration DESC
LIMIT 1
'''
pd.read_csv(IO(queryCartoDB(task4_1)), sep=',')

the_geom_webmercator  cartodb_id  \
0  0101000020110F000020D6742CE16A5FC112F4D73D73FC...          42   

   start_station_name  average_trip_duration  number_trips  
0  Broadway & W 49 St            1010.104167            48

### Answer: Broadway & W 49 St is the station that had the longest average trip duration during weekends and within 500m of TimeSquare

## Extra: create lines for trips started from stations within 500m of Times Squares and lasted less than 2 hours. The number of trips per each pair of stations are output as attributes of these lines. 

In [21]:
task4_2 = '''
SELECT MIN(cartodb_id) as cartodb_id,
       MIN(start_station_name) as start_station_name,
       MIN(end_station_name) as end_station_name,
       COUNT(cartodb_id) as number_trips,
       AVG(tripduration) as average_trip_duration      
FROM citibike
WHERE ST_DWithin(CDB_LatLng(start_station_latitude, 
                    start_station_longitude)::geography,
                    CDB_LatLng(40.7577,-73.9857)::geography,500) 
      AND tripduration <= 7200
GROUP BY start_station_id, start_station_latitude, start_station_longitude, 
            end_station_id, end_station_latitude, end_station_longitude

'''
pd.read_csv(IO(queryCartoDB(task4_2)), sep=',')

cartodb_id  start_station_name            end_station_name  number_trips  \
0         18466  Broadway & W 49 St             2 Ave & E 31 St             2   
1          2226     W 41 St & 8 Ave             W 21 St & 6 Ave             2   
2         15621  Broadway & W 41 St       E 48 St & Madison Ave             2   
3         34734     W 45 St & 6 Ave             W 29 St & 9 Ave             1   
4         37006     W 43 St & 6 Ave            W 52 St & 11 Ave             1   
5          6007  Broadway & W 41 St             W 24 St & 7 Ave             3   
6         20272     W 43 St & 6 Ave     E 37 St & Lexington Ave             4   
7         11540     W 45 St & 8 Ave            W 20 St & 11 Ave             3   
8         13614     W 45 St & 6 Ave             W 27 St & 7 Ave             3   
9         39800     W 41 St & 8 Ave    Elizabeth St & Hester St             1   
10        29033  Broadway & W 41 St                   Cherry St             1   
11        45442  Broadway & W 49 St             W 45 St & 6 Ave             1   
12        14181     W 45 St & 6 Ave             E 51 St & 1 Ave             2   
13          959     W 45 St & 8 Ave            W 47 St & 10 Ave             4   
14         7889     W 42 St & 8 Ave             E 47 St & 2 Ave             2   
15         7076     W 45 St & 6 Ave             W 41 St & 8 Ave            12   
16        33533     W 45 St & 8 Ave             E 40 St & 5 Ave             1   
17        34712  Broadway & W 41 St             W 11 St & 6 Ave             1   
18        13230     W 42 St & 8 Ave            W 59 St & 10 Ave             5   
19         4507     W 41 St & 8 Ave       Greenwich Ave & 8 Ave             2   
20        13031     W 41 St & 8 Ave    E 43 St & Vanderbilt Ave            31   
21        17712     W 43 St & 6 Ave            W 43 St & 10 Ave             1   
22        34501     W 45 St & 6 Ave            11 Ave & W 41 St             1   
23         9844     W 41 St & 8 Ave          Broadway & W 41 St             2   
24          356     W 41 St & 8 Ave             W 33 St & 7 Ave             5   
25        10223  Broadway & W 41 St          Broadway & W 24 St             5   
26        25199     W 41 St & 8 Ave          Broadway & W 51 St             2   
27         2536     W 45 St & 8 Ave             W 51 St & 6 Ave             2   
28        37521     W 41 St & 8 Ave            10 Ave & W 28 St             1   
29        19530     W 41 St & 8 Ave          Broadway & W 29 St             4   
..          ...                 ...                         ...           ...   
542       22083     W 42 St & 8 Ave     E 53 St & Lexington Ave             3   
543       34099  Broadway & W 49 St            W 46 St & 11 Ave             2   
544        7354  Broadway & W 49 St     Lexington Ave & E 24 St             1   
545       16753     W 45 St & 6 Ave            W 43 St & 10 Ave             4   
546       21315  Broadway & W 49 St  South End Ave & Liberty St             1   
547       38470     W 43 St & 6 Ave            11 Ave & W 41 St             1   
548       30699     W 41 St & 8 Ave       Warren St & Church St             1   
549       12092  Broadway & W 41 St             E 16 St & 5 Ave             3   
550       12625     W 45 St & 6 Ave          Broadway & W 58 St             1   
551       22819     W 45 St & 8 Ave     E 55 St & Lexington Ave             1   
552        8767  Broadway & W 41 St             E 33 St & 2 Ave             3   
553       14978     W 45 St & 6 Ave             W 45 St & 8 Ave             3   
554        7686  Broadway & W 41 St             E 20 St & 2 Ave             6   
555        6699     W 42 St & 8 Ave             W 45 St & 6 Ave             1   
556       17638     W 41 St & 8 Ave         Washington Square E             1   
557       28510     W 43 St & 6 Ave             W 27 St & 7 Ave             2   
558       17023     W 45 St & 6 Ave          E 14 St & Avenue B             1   
559        3105     W 45 St & 6 Ave 

## ASH Score 100. Extra Credit 100

# extra credit: make the function python 2 and 3 compatible so that it works on the  PUI2016_Python3 kernel